noria-server/mir/src/optimize.rs

```diff
@@ -1,19 +1,217 @@
+
+                                // But wait -- need to check if the filter is on the aggregation result
+                                use nom_sql::FunctionExpression::{Count, Sum};
+                                for (i, col) in child.columns.iter().enumerate() {
+                                    match col.function.clone() {
+                                        Some(func_expr) => {
+                                            match *func_expr {
+                                                 Count(col, _)
+                                                 | Sum(col, _) => {
+                                                    if col.name == on.name {
+                                                        // this column may be the aggregation result
+                                                        // so if we're filtering on it, we're not a candidate
+                                                        for (j, _cond) in conditions {
+                                                            if *j == i {
+                                                                candidate = false;
+                                                            }
+                                                        }
+                                                    }
+                                                },
+                                                _ => {},
+                                            }
+                                        },
+                                        None => {},
+                                    }
+                                }
+                            }
+                        },
+                        _ => {},
+                    };
+                }
+            },
+            _ => {},
+        };
+        if candidate {
+            candidate_nodes.push(n);
+        }
+    }
+
+    // 3. For each candidate, merge it, and update all parents/children
+    // of the newly merged node.
+
+    let mut new_nodes = Vec::new();
+
+    for n in candidate_nodes {
+        let temp = n.borrow();
+        let child = temp.children.first().unwrap().borrow();
+        // determine which is which
+        // (this was relevant when we supported either order of agg/filter,
+        // but I'm leaving it in place in case we resume that)
+        let (agg, filter) = match n.borrow().inner {
+            MirNodeType::Aggregation { .. }  => (n.borrow().inner.clone(), child.inner.clone()),
+            MirNodeType::Filter { .. } => (child.inner.clone(), n.borrow().inner.clone()),
+            _ => unreachable!(),
+        };
+
+        let mut new_name = child.name.clone();
+        new_name.push_str("_filteragg");
+
+        let new_node = MirNode::new(
+            &new_name,
+            child.from_version,
+            child.columns.clone(),
+            MirNodeType::FilterAggregation {
+                on: match &agg {
```

Comment:
This should probably be a `if let`.

Reply:
Hmm, I do like having the unreachable() to cause an assertion failure if it ever does end up hitting that code path. Does it work if I just use "let" instead of "if let"? 

(Mostly musing out loud; I can check for myself whether that compiles.)

Reply:
Sure!
```rust
if let MirNodeType::Aggregation { ... } = agg {
    // ...
} else {
    unreachable!()
}
```

Reply:
Yeah, I ended up with that, but I wouldn't call it easily readable :P